tests/catch/unit/stream/hipStreamWaitEvent.cc

```diff
@@ -0,0 +1,131 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+
+#include <hip_test_common.hh>
+
+TEST_CASE("Unit_hipStreamWaitEvent_Negative") {
+  enum class StreamTestType { NullStream = 0, StreamPerThread, CreatedStream };
+
+  auto streamType = GENERATE(StreamTestType::NullStream, StreamTestType::StreamPerThread,
+                             StreamTestType::CreatedStream);
+
+  hipStream_t stream{nullptr};
+  hipEvent_t event{nullptr};
+
+  if (streamType == StreamTestType::StreamPerThread) {
+    stream = hipStreamPerThread;
+  } else if (streamType == StreamTestType::CreatedStream) {
+    HIP_CHECK(hipStreamCreate(&stream));
+  }
+
+  HIP_CHECK(hipEventCreate(&event));
+
+  REQUIRE((stream != nullptr) != (streamType == StreamTestType::NullStream));
+  REQUIRE(event != nullptr);
+
+  SECTION("Invalid Event") {
+    INFO("Running against Invalid Event");
+    HIP_CHECK_ERROR(hipStreamWaitEvent(stream, nullptr, 0), hipErrorInvalidResourceHandle);
+  }
+
+  SECTION("Invalid Flags") {
+    INFO("Running against Invalid Flags");
+    constexpr unsigned flag = ~0u;
+    REQUIRE(flag != 0);
+    HIP_CHECK_ERROR(hipStreamWaitEvent(stream, event, flag), hipErrorInvalidValue);
+  }
+
+  HIP_CHECK(hipEventDestroy(event));
+
+  if (streamType == StreamTestType::CreatedStream) {
+    HIP_CHECK(hipStreamDestroy(stream));
+  }
+}
+
+// Kernel that runs a lambda which is passed on to it, with any arguments, thats passed along side
+// lambda.
+template <typename F, typename... Args> __global__ void launchFunctor(F f, Args... args) {
+  f(args...);
+}
+
+TEST_CASE("Unit_hipStreamWaitEvent_DifferentStreams") {
+  hipStream_t blockedStreamA{nullptr}, streamBlockedOnStreamA{nullptr}, unblockingStream{nullptr};
+  hipEvent_t waitEvent{nullptr};
+
+  HIP_CHECK(hipStreamCreate(&blockedStreamA));
+  HIP_CHECK(hipStreamCreate(&streamBlockedOnStreamA));
+  HIP_CHECK(hipStreamCreate(&unblockingStream));
+  HIP_CHECK(hipEventCreate(&waitEvent));
+
+  REQUIRE(blockedStreamA != nullptr);
+  REQUIRE(streamBlockedOnStreamA != nullptr);
+  REQUIRE(waitEvent != nullptr);
+
+  auto setValLambda = [] __device__(int* ptr, int val) { *ptr = val; };
+  auto waitLambda = [] __device__(int* sema) {
+    while (atomicCAS(sema, 1, 0) == 0)  //  CAS 1 with 0, basically unset it
+      ;
+  };
+
+  int *d_a{nullptr}, *waitData{nullptr};
+  HIP_CHECK(hipMalloc(&d_a, sizeof(int)));
+  HIP_CHECK(hipMalloc(&waitData, sizeof(int)));
+
+  HIP_CHECK(hipMemset(d_a, 0, sizeof(int)));
+  HIP_CHECK(hipMemset(waitData, 0, sizeof(int)));
+
+  launchFunctor<<<1, 1, 0, blockedStreamA>>>(setValLambda, d_a, 1);  // d_a = 1
+  launchFunctor<<<1, 1, 0, blockedStreamA>>>(waitLambda,
```

Comment:
Instead of using a different kernel to unblock the other one, use host allocation to do that. This way you cannot run into a deadlock in case resources are scarce or if context switch isn't enabled.

Reply:
any update on this?

Reply:
Working on a different way to test wait events.